### DISCUSSION TITLE
Strange behavior No such file or directory when run from file

### DISCUSSION THREAD

**Discussion by Angelk90** [zed-industries/zed#22059] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

<img width="849" alt="Screenshot 2024-12-06 alle 15 38 35" src="https://github.com/user-attachments/assets/d51c17b9-0880-487e-a785-fa8636c44bda">

When run from file by playing the button next to main:
```shell
Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.10s
Running `/Users/USERNAME/Desktop/file/rust_docx/target/debug/main3`
Error: Os { code: 2, kind: NotFound, message: "No such file or directory" }
```

When run from terminal via command `cargo run --bin main3`:
```shell
Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.12s
Running `target/debug/main3`
[{"field_name": "name", "type": "text", "default": ""}]
```
Then the file is found and read correctly.

The file is located one level up from the open repository.

Why does this happen?

Code main3.rs:

```rust
use std::fs::File;
use std::io::{self, Read};

fn main() -> io::Result<()> {
    let mut file = File::open("../main.txt")?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;
    println!("{}", contents);
    Ok(())
}
```

Cargo.toml
```cargo
....
[[bin]]
name = "main3"
path = "src/main3.rs"
```

### Environment

```
Zed: v0.165.2 (Zed Preview)
OS: macOS 14.1.1
Memory: 16 GiB
Architecture: x86_64
```

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary>

```

```
</details>

**Comment by RemcoSmitsDev** [comment]
See https://github.com/zed-industries/zed/issues/20873#issuecomment-2491811081

**Comment by Angelk90** [comment]
@RemcoSmitsDev : 

<img width="872" alt="Screenshot 2024-12-07 alle 17 51 56" src="https://github.com/user-attachments/assets/8e3c9d04-0677-4a68-a772-8983664fbd45">


Result play terminal:

```
Error: Os { code: 2, kind: NotFound, message: "No such file or directory" }
```

Am I doing something wrong?

**Comment by RemcoSmitsDev** [comment]
Your main.txt lives outside your project root, right? Could you test that if you print out the **CWD** you get the project root directory?

**Comment by Angelk90** [comment]
@RemcoSmitsDev : yes, one level higher.

Using `CWD` on the terminal I get:
`zsh: command not found: CWD
`

If I use pwd: `/Users/USERNAME/Desktop/file/rust_docx`

**Comment by RemcoSmitsDev** [comment]
Can you print out `env::current_dir()` inside your main function? It should be the same as `/Users/USERNAME/Desktop/file/rust_docx`.

**Comment by Angelk90** [comment]
@RemcoSmitsDev : I don't understand now it's working.

```
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/main3`
The current directory is: /Users/USERNAME/Desktop/file/rust_docx
[{"field_name": "name", "type": "text", "default": ""}]

⏵ Task `cargo run -p rust_docx --bin main3` finished successfully
⏵ Command: /bin/zsh -i -c 'cargo run -p rust_docx --bin main3'
```

Code:

```
use std::env;
use std::fs::File;
use std::io::{self, Read};

fn main() -> io::Result<()> {
    match env::current_dir() {
        Ok(path) => println!("The current directory is: {}", path.display()),
        Err(e) => println!("Error retrieving the current directory: {}", e),
    }

    let mut file = File::open("../main.txt")?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;
    println!("{}", contents);
    Ok(())
}
```

The reason it works seems that after adding the `.zed/tasks.json` file, you have to restart the project, so I closed `Zed` and then it seems to work.
In fact if there was the `.zed/tasks.json` file before it works, if I rename the `tasks.json` file, in fact it continues to work, until I restart `Zed`.
In fact restarting `Zed`, with the renamed file, trying again it doesn't work anymore.

**Comment by RemcoSmitsDev** [comment]
Ahh, great to hear!

**Comment by Angelk90** [comment]
@RemcoSmitsDev : But the fact of having to restart Zed, I don't understand why this happens.

There is a way to prevent these texts from appearing:

<img width="519" alt="Screenshot 2024-12-07 alle 18 28 40" src="https://github.com/user-attachments/assets/1eb8c5a4-d052-44fe-8c54-4ade4194ab68">

**Comment by RemcoSmitsDev** [comment]
Yeah, might be that the tasks are cached based on the `label` key. But I'm not sure about that, but that is a different issue you might want to create an issue for that and close the current one.

**Comment by Angelk90** [comment]
@RemcoSmitsDev : 
1) There is a way to prevent these texts from appearing:

<img width="519" alt="Screenshot 2024-12-07 alle 18 28 40" src="https://github.com/user-attachments/assets/1eb8c5a4-d052-44fe-8c54-4ade4194ab68">

2) You could open the problem, I think you could explain the problem better than me.

3) The difference I noticed that:

Not using `tasks.json` file running is executed on the full path:
```
Running `/Users/USERNAME/Desktop/file/rust_docx/target/debug/main3`
```
Using `tasks.json` file running is executed on the partial path:
```
Running `target/debug/main3`
```

**Comment by Angelk90** [comment]
@RemcoSmitsDev : 

https://github.com/user-attachments/assets/0d164a40-7222-4b2a-a977-c686217d4c1a

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
